python/ray/experimental/channel/shared_memory_channel.py

```diff
@@ -480,6 +484,71 @@ def close(self) -> None:
         self._worker.core_worker.experimental_channel_set_error(self._reader_ref)
 
 
+@DeveloperAPI
+class BufferedSharedMemoryChannel(ChannelInterface):
+    """A Channel that allow to buffer inputs.
+
+    The write/read APIs are non-blocking as long as the buffers are not
+    exhausted. The caller of the API should guarantee to consume buffers
+    before they are exhausted, otherwise write/read raises
+    `RayChannelTimeoutError`.
+    """
+
+    def __init__(
+        self,
+        writer: Optional[ray.actor.ActorHandle],
+        reader_and_node_list: List[Tuple["ray.actor.ActorHandle", str]],
+        num_shm_buffers: int,
+        typ: Optional[Union[int, SharedMemoryType]] = None,
+    ):
+        self._num_shm_buffers = num_shm_buffers
+        self._channels = [
+            Channel(writer, reader_and_node_list, typ) for _ in range(num_shm_buffers)
+        ]
+        self._next_write_index = 0
```

Comment:
Can you add some comments to explain how `_next_write_index` works? Maybe we could use `self.execution_index` instead and increment it whenever write or read is called on the writer and reader. When retrieving the buffer, we can use `execution_index % num_shm_buffers`. Using `execution_index` will also make debugging easier.

Reply:
Are you saying we should pass `self.execution_index` from a caller?